keps/sig-architecture/2558-field-version-openapi/README.md

```diff
@@ -0,0 +1,414 @@
+- Publish the default status of a feature gate (enabled/disabled by default)
+in OpenAPI.
+- Allow parsing comment tags on feature gates in [`kube_features.go`].
+- Update the Kubernetes API reference docs to include the field-level
+versioning information (out of scope of this KEP, can be done in a follow up).
+
+## Proposal
+
+This KEP proposes changes to the kube-openapi generator to add a new
+vendor extension field `x-kubernetes-field-version` in OpenAPI.
+This extension specifies the prerelease status of an API field,
+the minimum Kubernetes version for the prerelease and the
+feature gate associated with the field.
+
+This information is derived from comment tags on API fields with the format:
+
+```
+// +k8s:openapi-gen:prerelease=<prerelease-status>,minVersion=<k8s-version>,featureGate=<featuregate-name>
+```
+
+- The `prerelease` tag must be one of `alpha`, `beta`, `stable`
+or `deprecated`.
+- The `minVersion` tag denotes that the field is in the prerelease
+since the specified Kubernetes version. It must map to a Kubernetes
+release (e.g. `v1.20`).
+- The `featureGate` tag must be a valid feature gate as defined in [`kube_features.go`].
+
+The comment tag will be removed when the feature gate for the API field is removed.
+This will remove the `x-kubernetes-field-version` vendor extension for
+the field in OpenAPI.
+
+<<[UNRESOLVED]>>
+
+Open questions from @thockin's proposal:
+
+1) Should the tooling derive the prerelease status by looking up the gate name
+(one place to assert).
+
+2) Should we also tag structs or even new Kinds when they are gated?  This is
+less common and less well-defined.
+
+3) Should we tag *all* fields, even those that are not gated, so that tooling
+can know that the absence of the tag is an error, rather than indicating GA.
+
+<<[/UNRESOLVED]>>
+
+### Example
+
+Consider adding a new alpha API field `Width` to the stable API `Frobber`
+in `v1.20`.
+
+```go
+type Frobber struct {
+  Height *int32 `json:"height"
+  Param  string `json:"param"
+  // width indicates how wide the object is.
+  // This field is alpha-level and is only honored by servers that enable the Frobber2D feature.
+  // +optional
+  Width  *int32 `json:"width,omitempty"
+}
+```
+
+The field `Width` is controlled by the `Frobber2D` feature gate defined
+in [`kube_features.go`].
+
+```go
+// owner: @you
+// alpha: v1.20
+//
+// Add multiple dimensions to frobbers.
+Frobber2D utilfeature.Feature = "Frobber2D"
+
+var defaultKubernetesFeatureGates = map[utilfeature.Feature]utilfeature.FeatureSpec{
+  ...
+  Frobber2D: {Default: false, PreRelease: utilfeature.Alpha},
+}
+```
+
+To link the feature gate with the API field, add the following comment line
+with the `+k8s:openapi-gen` prefix for the `Width` field:
+
+```go
+type Frobber struct {
+  Height *int32 `json:"height"
+  Param  string `json:"param"
+  // width indicates how wide the object is.
+  // +optional
+  // +k8s:openapi-gen:prerelease=alpha,minVersion=v1.20,featureGate=Frobber2D
```

Comment:
Not sure about:

- all new fields will be gated by a feature gate? what about exceptions?
- what is the mapping between a field-level prerelease status and that of the feature gate? we will have a Beta feature gate guarding an Alpha field?

Reply:
> all new fields will be gated by a feature gate? 

All new fields in an existing API version are treated as alpha fields - https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#alpha-field-in-existing-api-version

> what is the mapping between a field-level prerelease status and that of the feature gate? we will have a Beta feature gate guarding an Alpha field?

IIUC the prerelease status of a field is denoted by the feature gate i.e. a field is alpha if the feature gate controlling the field is alpha.

cc @liggitt